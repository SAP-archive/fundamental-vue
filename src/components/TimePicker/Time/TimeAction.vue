<template>
  <div class="fd-time__control">
    <Button
      :state="buttonState"
      :icon="icon"
      compact
      styling="light"
      :type="type"
      :aria-label="ariaLabel"
      :aria-controls="ariaControls"
      @click="$emit('click')"
    >
      {{ title }}
    </Button>
  </div>
</template>

<script>
import ButtonTypes from "./../../Button/ButtonTypes";
import { Button } from "./../../Button";

const isButtonType = value => ButtonTypes.indexOf(value) >= 0;
export default {
  name: "FdTimeAction",
  components: { Button },
  computed: {
    buttonState() {
      return this.disabled ? "disabled" : "normal";
    }
  },
  props: {
    title: String,
    icon: String,
    type: { type: String, validator: isButtonType },
    disabled: Boolean,
    ariaLabel: String,
    ariaControls: String
  }
};
</script>
