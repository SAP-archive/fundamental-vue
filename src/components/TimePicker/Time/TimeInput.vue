<template>
  <div class="fd-time__input">
    <input
      @input="$emit('update', $event.target.value)"
      class="fd-form__control"
      maxlength="2"
      type="text"
      :value="value"
      v-bind="$attrs"
      v-on="$listeners"
    />
  </div>
</template>

<script>
export default {
  inheritAttrs: false,
  name: "FdTimeInput",
  props: ["value"]
};
</script>
