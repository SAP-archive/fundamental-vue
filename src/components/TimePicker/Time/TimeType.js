// Time type
const typeMapping = {
  hour24: "hour24",
  hour12: "hour12",
  minute: "minute",
  second: "second",
  meridian: "meridian"
};
export const TimeTypeList = Object.keys(typeMapping);
export const isTimeType = value => TimeTypeList.indexOf(value) >= 0;
