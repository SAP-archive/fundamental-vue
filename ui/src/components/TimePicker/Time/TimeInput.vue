<template>
  <div class="fd-time__input">
    <input
      @input="$emit('update', $event.target.value)"
      class="fd-form__control"
      maxlength="2"
      type="text"
      :value="value"
      v-bind="$attrs"
      v-on="$listeners"
    />
  </div>
</template>

<script lang="ts">
import Vue from "vue";

export default Vue.extend({
  inheritAttrs: false,
  name: "FdTimeInput",
  props: ["value"]
});
</script>
