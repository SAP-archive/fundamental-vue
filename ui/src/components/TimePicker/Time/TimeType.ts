// Time type
const typeMapping = {
  hour24: "hour24",
  hour12: "hour12",
  minute: "minute",
  second: "second",
  meridian: "meridian"
};
export type TimeType = keyof (typeof typeMapping);
export const TimeTypeList = Object.keys(typeMapping) as TimeType[];
export const isTimeType = (value: any) => TimeTypeList.indexOf(value) >= 0;
