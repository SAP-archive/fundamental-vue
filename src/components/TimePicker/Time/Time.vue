<template>
  <div class="fd-time__item" :aria-label="ariaLabel">
    <TimeAction icon="navigation-up-arrow" type="standard" @click="increaseValue" />
    <TimeInput :value="sanitizeInputValue" @update="timeUpdate" :placeholder="placeholder" />
    <TimeAction icon="navigation-down-arrow" type="standard" @click="decreaseValue" />
  </div>
</template>

<script>
import TimeRange from "../mixins/TimeRange";
import TimeAction from "./TimeAction.vue";
import TimeInput from "./TimeInput.vue";
import { isTimeType } from "./TimeType";

export default {
  name: "FdTime",
  mixins: [TimeRange],
  components: { TimeAction, TimeInput },
  props: {
    type: {
      type: String,
      required: true,
      validator: isTimeType
    },
    value: {
      type: [String, Number],
      default: ""
    },
    ariaLabel: {
      type: String,
      default: "Time Item"
    },
    placeholder: {
      type: String,
      default: ""
    }
  },
  computed: {
    sanitizeInputValue() {
      let value;
      let isValInRange;
      if (
        this.type === "hour24" ||
        this.type === "hour12" ||
        this.type === "minute" ||
        this.type === "second"
      ) {
        isValInRange = this.checkValueRange(this.inputValue, this.type);
        value = isValInRange ? this.inputValue : "";
      } else if (this.type === "meridian") {
        const meridian = this.inputValue.toString().toLowerCase();
        value =
          meridian === this.range[this.type].min || meridian === this.range[this.type].max
            ? meridian
            : "";
      } else {
        value = "";
      }
      return value;
    },
    previousValue() {
      const inputValue = Number(this.inputValue);
      const range = this.range[this.type];
      const { min, max } = range;
      const previousValue = inputValue > Number(min) ? inputValue - 1 : max;
      return previousValue.toString().padStart(2, "0");
    },
    nextValue() {
      const inputValue = Number(this.inputValue);
      const range = this.range[this.type];
      const { min, max } = range;
      const nextValue = inputValue < Number(max) ? inputValue + 1 : min;
      return nextValue.toString().padStart(2, "0");
    }
  },
  watch: {
    value: {
      immediate: true,
      handler(newValue) {
        this.inputValue = newValue;
        this.$emit("input", this.inputValue);
      }
    }
  },
  methods: {
    decreaseValue() {
      let value;
      let isValInRange;
      if (
        this.type === "hour24" ||
        this.type === "hour12" ||
        this.type === "minute" ||
        this.type === "second"
      ) {
        value = !isNaN(Number(this.inputValue)) ? this.previousValue : this.range[this.type].min;
        isValInRange = this.checkValueRange(value, this.type);
        value = isValInRange ? value : this.range[this.type].min;
      } else if (this.type === "meridian") {
        value =
          this.inputValue.toString().toLowerCase() === this.range[this.type].min
            ? this.range[this.type].max
            : this.range[this.type].min;
      } else {
        value = "";
      }
      this.inputValue = value;
      this.$emit("update:value", this.inputValue);
      this.emitTimeUpdate();
    },
    increaseValue() {
      let value;
      if (
        this.type === "hour24" ||
        this.type === "hour12" ||
        this.type === "minute" ||
        this.type === "second"
      ) {
        value = !isNaN(Number(this.inputValue)) ? this.nextValue : this.range[this.type].min;
      } else if (this.type === "meridian") {
        value =
          this.inputValue.toString().toLowerCase() === this.range[this.type].min
            ? this.range[this.type].max
            : this.range[this.type].min;
      } else {
        value = "";
      }
      this.inputValue = value;
      this.$emit("update:value", this.inputValue);
      this.emitTimeUpdate();
    },
    emitTimeUpdate() {
      this.$emit("timeUpdate", this.inputValue);
    },
    sanitizeValue() {
      let value;
      let isValInRange;
      if (
        this.type === "hour24" ||
        this.type === "hour12" ||
        this.type === "minute" ||
        this.type === "second"
      ) {
        isValInRange = this.checkValueRange(this.inputValue, this.type);
        value = isValInRange ? this.inputValue : "";
      } else if (this.type === "meridian") {
        const meridian = this.inputValue.toString().toLowerCase();
        value =
          meridian === this.range[this.type].min || meridian === this.range[this.type].max
            ? meridian
            : "";
      } else {
        value = "";
      }
      this.inputValue = value;
      this.$emit("update:value", this.inputValue);
    },
    timeUpdate(newValue) {
      if (
        this.type === "meridian" &&
        (newValue !== this.range[this.type].min && newValue !== this.range[this.type].max)
      ) {
        this.inputValue = "";
      } else {
        this.inputValue = newValue;
      }
      this.emitTimeUpdate();
    }
  },
  data() {
    return {
      inputValue: this.value
    };
  }
};
</script>
