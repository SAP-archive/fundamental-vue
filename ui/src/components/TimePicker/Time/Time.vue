<template>
  <div class="fd-time__item" :aria-label="ariaLabel">
    <TimeAction
      icon="navigation-up-arrow"
      type="standard"
      @click="increaseValue"
    />
    <TimeInput
      :value="sanitizeInputValue"
      @update="timeUpdate"
      :placeholder="placeholder"
    />
    <TimeAction
      icon="navigation-down-arrow"
      type="standard"
      @click="decreaseValue"
    />
  </div>
</template>

<script lang="ts">
import { PropValidator } from "vue/types/options";
import { mixins } from "@/mixins";
import TimeRange from "../mixins/TimeRange";
import TimeAction from "./TimeAction.vue";
import TimeInput from "./TimeInput.vue";
import { TimeType, isTimeType } from "./TimeType";

export default mixins(TimeRange).extend({
  name: "FdTime",
  components: { TimeAction, TimeInput },
  props: {
    type: {
      type: String,
      required: true,
      validator: isTimeType
    } as PropValidator<TimeType>,
    value: {
      type: [String, Number],
      default: ""
    },
    ariaLabel: {
      type: String,
      default: "Time Item"
    },
    placeholder: {
      type: String,
      default: ""
    }
  },
  computed: {
    sanitizeInputValue(): string | number {
      let value: string | number;
      let isValInRange: boolean;
      if (
        this.type === "hour24" ||
        this.type === "hour12" ||
        this.type === "minute" ||
        this.type === "second"
      ) {
        isValInRange = this.checkValueRange(this.inputValue, this.type);
        value = isValInRange ? this.inputValue : "";
      } else if (this.type === "meridian") {
        const meridian = this.inputValue.toString().toLowerCase();
        value =
          meridian === this.range[this.type].min ||
          meridian === this.range[this.type].max
            ? meridian
            : "";
      } else {
        value = "";
      }
      return value;
    },
    previousValue(): string {
      const inputValue = Number(this.inputValue);
      const range = this.range[this.type];
      const { min, max } = range;
      const previousValue = inputValue > Number(min) ? inputValue - 1 : max;
      return previousValue.toString().padStart(2, "0");
    },
    nextValue(): string {
      const inputValue = Number(this.inputValue);
      const range = this.range[this.type];
      const { min, max } = range;
      const nextValue = inputValue < Number(max) ? inputValue + 1 : min;
      return nextValue.toString().padStart(2, "0");
    }
  },
  watch: {
    value: {
      immediate: true,
      handler(newValue: string) {
        this.inputValue = newValue;
        this.$emit("input", this.inputValue);
      }
    }
  },
  methods: {
    decreaseValue() {
      let value: string;
      let isValInRange: boolean;
      if (
        this.type === "hour24" ||
        this.type === "hour12" ||
        this.type === "minute" ||
        this.type === "second"
      ) {
        value = !isNaN(Number(this.inputValue))
          ? this.previousValue
          : this.range[this.type].min;
        isValInRange = this.checkValueRange(value, this.type);
        value = isValInRange ? value : this.range[this.type].min;
      } else if (this.type === "meridian") {
        value =
          this.inputValue.toString().toLowerCase() === this.range[this.type].min
            ? this.range[this.type].max
            : this.range[this.type].min;
      } else {
        value = "";
      }
      this.inputValue = value;
      this.$emit("update:value", this.inputValue);
      this.emitTimeUpdate();
    },
    increaseValue() {
      let value: string;
      if (
        this.type === "hour24" ||
        this.type === "hour12" ||
        this.type === "minute" ||
        this.type === "second"
      ) {
        value = !isNaN(Number(this.inputValue))
          ? this.nextValue
          : this.range[this.type].min;
      } else if (this.type === "meridian") {
        value =
          this.inputValue.toString().toLowerCase() === this.range[this.type].min
            ? this.range[this.type].max
            : this.range[this.type].min;
      } else {
        value = "";
      }
      this.inputValue = value;
      this.$emit("update:value", this.inputValue);
      this.emitTimeUpdate();
    },
    emitTimeUpdate() {
      this.$emit("timeUpdate", this.inputValue);
    },
    sanitizeValue() {
      let value: string | number;
      let isValInRange: boolean;
      if (
        this.type === "hour24" ||
        this.type === "hour12" ||
        this.type === "minute" ||
        this.type === "second"
      ) {
        isValInRange = this.checkValueRange(this.inputValue, this.type);
        value = isValInRange ? this.inputValue : "";
      } else if (this.type === "meridian") {
        const meridian = this.inputValue.toString().toLowerCase();
        value =
          meridian === this.range[this.type].min ||
          meridian === this.range[this.type].max
            ? meridian
            : "";
      } else {
        value = "";
      }
      this.inputValue = value;
      this.$emit("update:value", this.inputValue);
    },
    timeUpdate(newValue: string) {
      if (
        this.type === "meridian" &&
        (newValue !== this.range[this.type].min &&
          newValue !== this.range[this.type].max)
      ) {
        this.inputValue = "";
      } else {
        this.inputValue = newValue;
      }
      this.emitTimeUpdate();
    }
  },
  data() {
    return {
      inputValue: this.value as string | number
    };
  }
});
</script>
