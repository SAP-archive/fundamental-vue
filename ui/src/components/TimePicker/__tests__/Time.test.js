import { mount } from "@vue/test-utils";
import Time from "../Time/Time.vue";

describe("Time", () => {
  let value = "23";
  const placeholder = "Hour";
  const ariaLabel = "24-Hour";

  const wrapper = mount(Time, {
    propsData: {
      type: "hour24",
      placeholder,
      ariaLabel,
      value
    }
  });

  const input = wrapper.find("input");
  expect(input).toBeDefined();

  it("renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders correct placeholder and aria-label", () => {
    expect(input.attributes("placeholder")).toBe(placeholder);
    expect(wrapper.find(".fd-time__item").attributes("aria-label")).toBe(
      ariaLabel
    );
  });

  it("emits events and has correct inputValue and input element value", () => {
    const navDown = wrapper.find(".sap-icon--navigation-down-arrow");
    navDown.trigger("click");

    const expected = "22";

    expect(wrapper.emitted("input").length).toBeGreaterThan(0);

    const timeUpdate = wrapper.emitted("timeUpdate");
    expect(timeUpdate.length).toBeGreaterThan(0);
    expect(timeUpdate[timeUpdate.length - 1][0]).toEqual(expected);

    expect(wrapper.vm.$data.inputValue).toEqual(expected);
    expect(input.element.value).toEqual(expected);
  });

  describe("Hour24", () => {
    const wrapper = mount(Time, {
      propsData: {
        type: "hour24",
        value: "23"
      }
    });

    it("navigates to correct value", () => {
      const navUp = wrapper.find(".sap-icon--navigation-up-arrow");
      const navDown = wrapper.find(".sap-icon--navigation-down-arrow");

      navUp.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("00");

      navUp.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("01");

      navDown.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("00");

      navDown.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("23");
    });
  });

  describe("Hour12", () => {
    const wrapper = mount(Time, {
      propsData: {
        type: "hour12",
        value: "11"
      }
    });

    it("navigates to correct value", () => {
      const navUp = wrapper.find(".sap-icon--navigation-up-arrow");
      const navDown = wrapper.find(".sap-icon--navigation-down-arrow");

      navUp.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("12");

      navUp.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("01");

      navDown.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("12");

      navDown.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("11");
    });
  });

  describe("Minute", () => {
    const wrapper = mount(Time, {
      propsData: {
        type: "minute",
        value: "01"
      }
    });

    it("navigates to correct value", () => {
      const navUp = wrapper.find(".sap-icon--navigation-up-arrow");
      const navDown = wrapper.find(".sap-icon--navigation-down-arrow");

      navDown.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("00");

      navDown.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("59");

      navUp.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("00");

      navUp.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("01");
    });
  });

  describe("Second", () => {
    const wrapper = mount(Time, {
      propsData: {
        type: "second",
        value: "58"
      }
    });

    it("navigates to correct value", () => {
      const navUp = wrapper.find(".sap-icon--navigation-up-arrow");
      const navDown = wrapper.find(".sap-icon--navigation-down-arrow");

      navUp.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("59");

      navUp.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("00");

      navDown.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("59");

      navDown.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("58");
    });
  });

  describe("Meridian", () => {
    const wrapper = mount(Time, {
      propsData: {
        type: "meridian",
        value: "am"
      }
    });

    it("navigates to correct value", () => {
      const navUp = wrapper.find(".sap-icon--navigation-up-arrow");
      const navDown = wrapper.find(".sap-icon--navigation-down-arrow");

      navUp.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("pm");

      navUp.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("am");

      navDown.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("pm");

      navDown.trigger("click");
      expect(wrapper.vm.$data.inputValue).toEqual("am");
    });
  });
});
