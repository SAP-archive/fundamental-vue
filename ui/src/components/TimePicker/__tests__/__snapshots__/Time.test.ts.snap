// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Time renders correctly 1`] = `
<div
  aria-label="24-Hour"
  class="fd-time__item"
>
  <div
    class="fd-time__control"
  >
    <button
      class="fd-button--compact fd-button--light fd-button--standard sap-icon--navigation-up-arrow"
    >
      
    </button>
  </div>
   
  <div
    class="fd-time__input"
  >
    <input
      class="fd-form__control"
      maxlength="2"
      placeholder="Hour"
      type="text"
    />
  </div>
   
  <div
    class="fd-time__control"
  >
    <button
      class="fd-button--compact fd-button--light fd-button--standard sap-icon--navigation-down-arrow"
    >
      
    </button>
  </div>
</div>
`;
